### PR TITLE
async grpc calls should always return void

### LINKED ISSUE #9257: Async grpc calls should always return void
### What is the problem?

The generated grpc client stubs always return a status. Even when the status returned by the stub indicates a failure (i.e., IOError), the callback that was passed in still gets fired. This is confusing because it can lead to handling a failure twice and the developer must be careful to always ignore the return value of the initial client call ([example](https://github.com/ray-project/ray/blob/e0d8b58969728054d1800abc925aa5352882baa8/src/ray/core_worker/transport/direct_task_transport.cc#L140)).

Instead, we should have the client stub return void.

### PR DESCRIPTION
**Pull request by zhijunfu**
## Why are these changes needed?

Refer to [this issue](https://github.com/ray-project/ray/issues/9257) for more details

## Related issue number

Closes #9257 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by rkooo567** [comment]
Nice change! Isn't service based accessor also supposed to follow the same pattern?

**Comment by zhijunfu** [comment]
> Nice change! Isn't service based accessor also supposed to follow the same pattern?

Not sure if I understand correctly. Do you see any additional places that need to be changed as well?

**Comment by rkooo567** [comment]
cc @stephanie-wang correct me if I am wrong. 

I think the initial motivation for this issue was that `service_based_accessor` rpc calls return Status although it only returns Status::OK(). And this PR seems to only handle core worker rpc calls, not the service_based_accessor.

**Comment by zhijunfu** [comment]
I think the original issue #9257 was talking about making async grpc call method return void, since the status has already been handled by the callback, while for gcs rpc client, all the methods defined in `gcs_rpc_client.h` are already void (all of them are defined as `VOID_GCS_RPC_CLIENT_METHOD`), I think they are already covered as far as grpc is concerned.

File: src/ray/rpc/grpc_client.h [hunk 1]
```diff
@@ -90,12 +83,12 @@ class GrpcClient {
...
   /// \return Status.
   template <class Request, class Reply>
-  ray::Status CallMethod(
+  void CallMethod(
       const PrepareAsyncFunction<GrpcService, Request, Reply> prepare_async_function,
       const Request &request, const ClientCallback<Reply> &callback) {
     auto call = client_call_manager_.CreateCall<GrpcService, Request, Reply>(
         *stub_, prepare_async_function, request, callback);
-    return call->GetStatus();
+    RAY_CHECK(call != nullptr);
```
Comment (ericl) [inline src/ray/rpc/grpc_client.h:91]: @zhijunfu Doesn't this throw away bad statuses? I was expecting something more like:

```
auto send = call->GetStatus();
if (!send.ok()) {
   callback(send, Reply());
}
```

**Review by ericl — CHANGES_REQUESTED** [review]
Comment inline

**Comment by rkooo567** [comment]
@zhijunfu I think all these service based accessor methods also return meaningless status https://github.com/ray-project/ray/blob/c290c308fe1e496480db5c37489df619cff6168f/src/ray/gcs/gcs_client/service_based_accessor.cc#L46. Maybe we should take care of it after deleting Redis Accessor in the future?

File: src/ray/rpc/grpc_client.h (hunk 1 above)
Comment (zhijunfu) [reply src/ray/rpc/grpc_client.h:91]: @ericl The assumption here is that the callback will be invoked by Gprc, no matter it succeeds or fails. With this assumption, the callback will be invoked twice if we call it here on failure as well.

**Comment by zhijunfu** [comment]
> @zhijunfu I think all these service based accessor methods also return meaningless status
> 
> https://github.com/ray-project/ray/blob/c290c308fe1e496480db5c37489df619cff6168f/src/ray/gcs/gcs_client/service_based_accessor.cc#L46
> 
> . Maybe we should take care of it after deleting Redis Accessor in the future?

Yep, this is kind of independent, and it would be easier to revisit this later after redis accessor is removed.

**Comment by ericl** [comment]
Are we sure this is true? Is it in the gRPC documentation or something

File: src/ray/rpc/grpc_client.h (hunk 1 above)
Comment (ericl) [reply src/ray/rpc/grpc_client.h:91]: @stephanie-wang is this true?

**Comment by zhijunfu** [comment]
There's a discussion in slack some days ago, I'm copying the content here below

@stephanie-wang previously mentioned the callback is always called per her observation. It's kind of implied in the links below, but I didn't find a place that explicitly said so. 

I think it's probably ok to make the change as you suggested, if it's fine for the callback to be invoked twice in some cases.

```
I did some research today, from the tutorial and examples I found for grpc, it’s probaby ok to assume the callback will be called, though I didn’t find something more definitive.
https://grpc.io/docs/languages/cpp/async/
https://github.com/grpc/grpc/blob/master/examples/cpp/helloworld/greeter_async_client2.cc
https://github.com/grpc/grpc/blob/master/test/cpp/qps/client_async.cc
```

> Are we sure this is true? Is it in the gRPC documentation or something

**Review by ericl — APPROVED** [review]
lgtm then. It would be a bit safer to audit gRPC to verify it is indeed always called, but I think we have been assuming this anyways already.

**Pull request merged**